text/0000-rustdoc-url-conflicts.md

```diff
@@ -0,0 +1,65 @@
+- Feature Name: rustdoc_url_conflicts
+- Start Date: 2021-03-29
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This RFC describes how the rustdoc URLs will be modified to fix the current issue on case insensitive file systems.
+For example, in the libstd there is:
+ * keyword.self.html
+ * keyword.Self.html
+
+If macOS or Windows users try to access to both URLs, they will always see the same content because for both file systems, it's the same file.
+
+# Motivation
+[motivation]: #motivation
+
+This is one of the oldest rustdoc issue and has been problematic for years and a lot of users.
```

Comment:
Could you provide more details about those places?

My first instinct here is that having two structs in the same module that differ only by case is something that people should never be doing, and thus this should never come up.  TBH I'd be inclined to make a deny-by-default lint for it, and say that if you do it and the docs don't work I don't really care.  But maybe there's an important situation I'm missing, so having that called out here could easily change my mind about it.

The places where things differ in only case that don't bother me are things like `fn.default.html` and `trait.Default.html`, where the item-kind disambiguator is already making the URL unique.

(Then we could special-case `keyword.*.html` to make it work.  Or just make one doc for both of them.)

Reply:
Yeah this seems like a _major_ change for very very little benefit in a case people should not be doing at all.

Reply:
But as soon as we enter into the "binding zone", it's just conflicts everywhere. What to do then? Rename items? That could work, but that doesn't sound like a good idea... This is one of the oldest rustdoc issue for a good reason after all. I think it's a really good change because it keeps a recognizable URL scheme while preventing name conflicts. But well, debates are what RFCs are for after all.

Reply:
Also, special case things is rarely a good idea.

Reply:
@GuillaumeGomez what do you mean by the binding zone?

I do not think using dashes randomly is a recognizeable URL scheme.

Reply:
Yes, by "randomly" I meant "not in a way that is intuitively predictable"

Reply:
> Also, special case things is rarely a good idea.

While I agree in general, I see no issue with special-casing `keyword`, since it's *already* a magic special case anyway.  It's not like users can add `keyword yolo;` items in their crates that we need to document.

Reply:
They actually can. :p

The idea behind `doc(keyword)` is to allow proc-macro crates to add documentation for their own keywords if they introduce new ones.